packages/devtools/test/integration_tests/server_test.dart

```diff
@@ -42,34 +59,180 @@ void main() {
+
+    // Ensure we now have a single connected client.
+    final serverResponse = await _waitForClients(requiredConnectionState: true);
+    expect(serverResponse['clients'], hasLength(1));
+    expect(serverResponse['clients'][0]['hasConnection'], isTrue);
+    expect(serverResponse['clients'][0]['vmServiceUri'],
+        equals(appFixture.serviceUri.toString()));
+
+    // TODO(dantup): This test will fail until the devtools pubspec.yaml
+    // references a version of devtools_server that has this support!
+  }, timeout: const Timeout.factor(10), skip: true);
+}
 
-    server.stderr.listen((text) => throw 'STDERR: $text');
+Future<void> _startApp() async {
+  appFixture = await CliAppFixture.create('test/fixtures/empty_app.dart');
 
-    server.write({
-      'id': '10',
-      'method': 'vm.register',
-      'params': {'uri': appFixture.serviceUri.toString()}
-    });
+  // TODO(dantup): When the stable versions of Dart + Flutter are >= v3.22
+  // of the VM Service (July 2019), the _Service option here can be removed.
+  final serviceStreamName =
+      await appFixture.serviceConnection.serviceStreamName;
 
-    // Expect to get a successful response from the server.
-    final serverResponse =
-        await server.output.firstWhere((e) => e['id'] == '10');
-    expect(serverResponse['result']['success'], isTrue);
+  // Track services method names as they're registered.
+  appFixture.serviceConnection
+      .onEvent(serviceStreamName)
+      .where((e) => e.kind == EventKind.kServiceRegistered)
+      .listen((e) => registeredServices[e.service] = e.method);
+  await appFixture.serviceConnection.streamListen(serviceStreamName);
+}
 
-    // Expect the VM service to see the launchDevTools service registered.
-    expect(registeredServices, contains('launchDevTools'));
-  }, timeout: const Timeout.factor(10));
+int nextId = 0;
+Future<Map<String, dynamic>> _send(String method,
+    [Map<String, dynamic> params]) {
+  final id = (nextId++).toString();
+  completers[id] = new Completer<Map<String, dynamic>>();
+  server.write({'id': id.toString(), 'method': method, 'params': params});
+  return completers[id].future;
+}
+
+// It may take time for the servers client list to be updated as the web app
+// connects, so this helper just polls waiting for the expected state and
+// then returns the client list.
+Future<Map<String, dynamic>> _waitForClients(
+    {bool requiredConnectionState}) async {
+  Map<String, dynamic> serverResponse;
+  await waitFor(
+    () async {
+      serverResponse = await _send('clients.list');
+      final clients = serverResponse['clients'];
+      return clients is List &&
+          clients.isNotEmpty &&
+          (requiredConnectionState == null ||
+              clients
+                  .any((c) => c['hasConnection'] == requiredConnectionState));
+    },
+    timeout: const Duration(seconds: 10),
+    timeoutMessage: 'Server did not return any known clients',
+    delay: const Duration(seconds: 1),
```

Comment:
using this sort of pattern instead of the 5 second timeout used earlier would also be fine.

Reply:
This one is easy because we're waiting for something to happen, but the other one has no visible effects (because it would reuse an existing connection, and the client list would look the same). I think an event will work though, so I'll do that.